### PR TITLE
Generalize storage system

### PR DESCRIPTION
**Pull request by olivershen-wow**
- Abstract interface StorageServiceClient for generalized storage implementation use;
- Refactor all usage of BlobStorageClient.java to StorageServiceClient.java;
- Add StorageManageService for StorageServiceClient init control;
- Enable fastjson auto type support for all entity classes under common/entity/common package, move classes that would be used in Websocket communication to common/entity/common;
- Modify CDNUrl usage in StorageFileInfo, from CDN endpoint to absolute CDN URL;
- Frontend compatibility, naming the token from BlobSignature to FileToken;
- Gradle plugin compatibility;
- Tested in plugin and frontend;
- Fix UT for modified classes;
- Fix: no photo-uri is configured when getting photo;
- Refactor entity class BlobFileInfo to StorageFileInfo, rename some columns of StorageFileInfo and TestJsonInfo:
    - Alter blobUrl -> fileDownloadUrl
    - Alter blobPath -> fileRelPath
    - Alter blobContainer -> storageContainer
- Alter center controller API:
    - /api/package/getSAS(Kept as deprecated, to be compatible to front-end/gradle plugin) -> /api/file/getToken 

Storage system class UML design diagram:
image
Storage system run sequence diagram:
image

**TODO:**
- **Local storage support;**
- **Update README with latest storage config in Docker images**
- **ADO plugin support;**
- **Before release: DB table update;**

### CODE REVIEW

**Comment by hydraxman** [comment]
THanks for the effort, could we split the PR into 2~3 changes to help balance the review cost?

File: common/src/main/java/com/microsoft/hydralab/common/file/impl/blob/BlobClientAdapter.java [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
-package com.microsoft.hydralab.common.util.blob;
+package com.microsoft.hydralab.common.file.impl.blob;
```
Comment (hydraxman) [inline common/src/main/java/com/microsoft/hydralab/common/file/impl/blob/BlobClientAdapter.java:3]:
```suggestion
package com.microsoft.hydralab.common.file.impl.azure;
```

Comment (olivershen-wow) [reply common/src/main/java/com/microsoft/hydralab/common/file/impl/blob/BlobClientAdapter.java]: Updated, thanks.

**Comment by olivershen-wow** [comment]
> THanks for the effort, could we split the PR into 2~3 changes to help balance the review cost? Plus, it would be helpful if we provide some diagram to describe the relations and interfaces.

@hydraxman Diagrams added in description. For the spliting stuff, the most changes are for refactoring and abstracting adapters. Only changes with no dependency are for front-end and gradle plugin. Only few files are changed, and are in separate modules, I think it's already easy to review now.

**Comment by hydraxman** [comment]
@olivershen-wow 
建议拆分PR为以下两部分：
- 对用户影响较大的 需要修改接入流程、代码的（gradle/接口参数refactor）独立一部分代码（这部分修改可能需要考虑向前兼容）
- 内部调整文件系统接口的，不影响用户接入流程的修改部分
如果这样做有挑战，随时和我说。目的是为了实现功能、减轻onboarding cost的基础上，尽可能少的影响用户。

File: agent/src/main/java/com/microsoft/hydralab/agent/service/AgentWebSocketClientService.java [hunk 2]
```diff
@@ -160,13 +162,19 @@ public void onMessage(Message message) {
...
-            log.info("Pushgateway has set basic auth now, data can be pushed correctly.");
+        prometheusPushgatewayInit(agentMetadata);
+    }
+
+    private void storageTypeConsistencyCheck(AccessToken accessToken) {
+        try {
+            storageServiceClient.storageTypeCheck(accessToken);
+        } catch (HydraLabRuntimeException e) {
+            log.error(e.getMessage());
+            System.exit(e.getCode());
```
Comment (hydraxman) [inline agent/src/main/java/com/microsoft/hydralab/agent/service/AgentWebSocketClientService.java:177]: Why exit the system here?

File: center/application-sample.yml [hunk 3]
```diff
@@ -15,6 +15,8 @@ spring:
 app:
   # Mail Address Format
   default-user: ${DEFAULT_USER}
-  blob:
-    connection: ${BLOB_CONNECTION_STR}
-    CDNUrl: ${CDN_URL}
\ No newline at end of file
+  storage:
+    type: AZURE # LOCAL, AZURE, ALICLOUD, etc.
```
Comment (hydraxman) [inline center/application-sample.yml:19]:
```suggestion
    type: AZURE # current available options: LOCAL, AZURE
```

File: agent/src/main/resources/application.yml [hunk 4]
```diff
@@ -55,6 +55,8 @@ app:
       count-threshold: ${STATE_CHANGE_COUNT_THRESHOLD:12}
       window-time: ${STATE_CHANGE_WINDOW_TIME:5}
       recovery-time: ${STATE_RECOVERY_TIME:3}
+  storage:
+    type: AZURE # LOCAL, AZURE, ALICLOUD, etc.
```
Comment (hydraxman) [inline agent/src/main/resources/application.yml:59]: Why agent side also need to config this? Could the agent just follow the center on this?

File: agent/src/test/java/com/microsoft/hydralab/agent/service/TestTaskEngineServiceTest.java [hunk 5]
```diff
@@ -4,7 +4,7 @@
 import com.microsoft.hydralab.agent.runner.TestRunner;
 import com.microsoft.hydralab.agent.runner.espresso.EspressoRunner;
 import com.microsoft.hydralab.agent.test.BaseTest;
-import com.microsoft.hydralab.common.entity.center.TestTaskSpec;
+import com.microsoft.hydralab.common.entity.common.TestTaskSpec;
```
Comment (hydraxman) [inline agent/src/test/java/com/microsoft/hydralab/agent/service/TestTaskEngineServiceTest.java:7]: Why do we need to move the spec to common?

**Review by hydraxman — COMMENTED** [review]
Thanks for the update, but still too many changes... If I want to do a careful review, I have to hit 90+ viewed checkboxes. please control the refactor changes

File: center/src/main/java/com/microsoft/hydralab/center/controller/StorageController.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * @author Li Shen
+ * @date 2/20/2023
+ */
+
+@RestController
+@RequestMapping
+// Used only for agent to push to / pull from local storage in Center, when no external storage is configured.
+public class StorageController {
```
Comment (hydraxman) [inline center/src/main/java/com/microsoft/hydralab/center/controller/StorageController.java:22]: Has this been used anywhere? If not, we may include this change in next PR maybe.

File: agent/src/main/java/com/microsoft/hydralab/agent/service/AgentWebSocketClientService.java (hunk 2 above)
Comment (olivershen-wow) [reply agent/src/main/java/com/microsoft/hydralab/agent/service/AgentWebSocketClientService.java]: Added StorageManageService for StorageServiceClient init, remove storage type check related code.

File: agent/src/main/java/com/microsoft/hydralab/agent/service/TestDataService.java [hunk 7]
```diff
@@ -2,10 +2,7 @@
 // Licensed under the MIT License.
 package com.microsoft.hydralab.agent.service;
 
-import com.microsoft.hydralab.common.entity.common.AndroidTestUnit;
-import com.microsoft.hydralab.common.entity.common.TestRun;
-import com.microsoft.hydralab.common.entity.common.EntityFileRelation;
-import com.microsoft.hydralab.common.entity.common.TestTask;
+import com.microsoft.hydralab.common.entity.common.*;
```
[the remark this answers is not in the document]
Comment (olivershen-wow) [reply agent/src/main/java/com/microsoft/hydralab/agent/service/TestDataService.java]: Resolved.

File: agent/src/main/resources/application.yml (hunk 4 above)
Comment (olivershen-wow) [reply agent/src/main/resources/application.yml]: Removed

File: center/application-sample.yml (hunk 3 above)
Comment (olivershen-wow) [reply center/application-sample.yml:19]: Resolved.

File: center/src/main/java/com/microsoft/hydralab/center/controller/StorageController.java (hunk 6 above)
Comment (olivershen-wow) [reply center/src/main/java/com/microsoft/hydralab/center/controller/StorageController.java]: Removed

File: common/src/main/java/com/microsoft/hydralab/common/util/StorageManageService.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+import org.springframework.context.ApplicationContext;
+
+import java.io.File;
+
+/**
+ * @author Li Shen
+ * @date 3/1/2023
+ */
+
+public class StorageManageService extends StorageServiceClient {
```
Comment (hydraxman) [inline common/src/main/java/com/microsoft/hydralab/common/util/StorageManageService.java:20]:
```suggestion
public class StorageServiceClientProxy extends StorageServiceClient {
```

File: center/src/main/resources/application.yml [hunk 9]
```diff
@@ -51,14 +51,18 @@ app:
...
-    fileLimitDay: 6
-    CDNUrl: ${CDN_URL:@null}
-    SASExpiryTimeFront: ${BLOB_SAS_EXPIRY_FRONT:120}
-    SASExpiryTimeAgent: ${BLOB_SAS_EXPIRY_AGENT:120}
-    SASExpiryUpdate: ${BLOB_SAS_EXPIRY_UPDATE:10}
-    timeUnit: ${TIME_UNIT:MINUTES}
+  storage:
+    type: AZURE # current available options: LOCAL, AZURE
+    blob:
+      connection: ${BLOB_CONNECTION_STR}
```
Comment (hydraxman) [inline center/src/main/resources/application.yml:57]: This change (change of the connection string name) has an impact on the service onboarding/out-of-box experience, you may need to update the related wiki as well.

File: center/src/main/java/com/microsoft/hydralab/center/interceptor/BaseInterceptor.java [hunk 10]
```diff
@@ -77,11 +77,11 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
...
             //invoke by browser
             if (StringUtils.isEmpty(token) || !authUtil.verifyToken(token)) {
-                if (requestURI.contains(Const.FontPath.PREFIX_PATH)) {
+                if (requestURI.contains(Const.FrontPath.PREFIX_PATH)) {
                     String queryString = request.getQueryString();
                     if (StringUtils.isNotEmpty(queryString)
-                            && queryString.startsWith(Const.FontPath.REDIRECT_PARAM)
-                            && LogUtils.isLegalStr(queryString.replace(Const.FontPath.REDIRECT_PARAM + "=", ""), Const.RegexString.URL, false)
+                            && queryString.startsWith(Const.FrontPath.REDIRECT_PARAM)
+                            && LogUtils.isLegalStr(queryString.replace(Const.FrontPath.REDIRECT_PARAM + "=", ""), Const.RegexString.URL, false)
```
Comment (hydraxman) [inline center/src/main/java/com/microsoft/hydralab/center/interceptor/BaseInterceptor.java:84]: Change this FontPath in another PR, this PR is already large.

File: center/src/main/resources/application-release.yml [hunk 11]
```diff
@@ -8,6 +8,7 @@ spring:
             authorization-uri: https://login.microsoftonline.com/${MICROSOFT_PROVIDER_TENANT_ID}/oauth2/v2.0/authorize
             token-uri: https://login.microsoftonline.com/${MICROSOFT_PROVIDER_TENANT_ID}/oauth2/v2.0/token
             jwk-set-uri: https://login.microsoftonline.com/${MICROSOFT_PROVIDER_TENANT_ID}/discovery/v2.0/keys
+            photo-uri: https://graph.microsoft.com/v1.0/me/photo/$value
```
Comment (hydraxman) [inline center/src/main/resources/application-release.yml:11]: Could you share why we need to add this uri?

File: common/src/main/java/com/microsoft/hydralab/common/file/ContainerConstants.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package com.microsoft.hydralab.common.file;
+
+public interface ContainerConstants {
```
Comment (hydraxman) [inline common/src/main/java/com/microsoft/hydralab/common/file/ContainerConstants.java:3]: If it's constants, shouldn't it only contain constants? If it's a getter collection, a better name is StorageProperties

File: common/src/main/java/com/microsoft/hydralab/common/file/impl/azure/AzureBlobProperty.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+public class AzureBlobProperty implements ContainerConstants {
+    private String connection;
+    private long SASExpiryTimeFront;
+    private long SASExpiryTimeAgent;
+    private long SASExpiryUpdate;
+    private String timeUnit;
+    private int fileLimitDay;
+    private String CDNUrl;
+
+    private String SCREENSHOT_CONTAINER_NAME = "images";
```
Comment (hydraxman) [inline common/src/main/java/com/microsoft/hydralab/common/file/impl/azure/AzureBlobProperty.java:26]:
```suggestion
    private final static String SCREENSHOT_CONTAINER_NAME = "images";
```
same applies to all below.

**Comment by hydraxman** [comment]
依然建议你新建一个分支和PR拆开这个改动，我们尽量一个PR做一件事，这样非常clean，而且你不用revert，把这个PR变成draft，用于存储所有改动，后续从里面摘取内容，拆成2-3个PR。

File: center/src/main/java/com/microsoft/hydralab/center/interceptor/BaseInterceptor.java (hunk 10 above)
Comment (olivershen-wow) [reply center/src/main/java/com/microsoft/hydralab/center/interceptor/BaseInterceptor.java]: Pruned, thx.

File: center/src/main/resources/application.yml (hunk 9 above)
Comment (olivershen-wow) [reply center/src/main/resources/application.yml:57]: For current changes, no extra/new operation is required for user of Uber image, the pass-in parameter is still BLOB_CONNECTION_STR.
I will update the README when local storage is implemented, I will illustrate all supported storage types of Hydra Lab and list the corresponding inputs then.

File: center/src/main/resources/application-release.yml (hunk 11 above)
Comment (olivershen-wow) [reply center/src/main/resources/application-release.yml:11]: It's a misdeleted uri for portal photo requesting, I've checked it in in another PR.

File: common/src/main/java/com/microsoft/hydralab/common/file/impl/azure/AzureBlobProperty.java (hunk 13 above)
Comment (olivershen-wow) [reply common/src/main/java/com/microsoft/hydralab/common/file/impl/azure/AzureBlobProperty.java]: Resolved.

File: common/src/main/java/com/microsoft/hydralab/common/file/ContainerConstants.java (hunk 12 above)
Comment (olivershen-wow) [reply common/src/main/java/com/microsoft/hydralab/common/file/ContainerConstants.java]: Resolved.

**Pull request closed without merging**